qutip/core/data/csc.pyx

```diff
@@ -185,6 +186,9 @@ cdef class CSC(base.Data):
         # be collected while we're alive.
         if self._scipy is not None:
             return self._scipy
+        if not self._deallocate:
+            # Does not own the data, thus cannot give it to scipy.
+            raise ValueError("Can't pass data ownership to scipy")
```

Comment:
We can't throw an exception in `as_scipy()` - it violates the core principle that Python space should be able to modify data-layer objects, which prevents users from being able to use the objects unless they compile their own code.

You can create `ndarray`s that don't own their data, but you need to set the `base` property of the `ndarray` to the relevant owner in Python space to ensure that the references don't go stale.  To do this, we'd need to add a `base` property to all the data-layer objects to store the ownership, and we're suddenly re-inventing numpy's refcounting, which we certainly will get wrong in the future.

As soon as we have views onto data, we should let numpy handle it.  The `as_scipy()` routines aren't very expensive at all, really, and having the only alternative to deallocation being "defer to numpy" rather than "defer to another data-layer object" means that we completely avoid having to do refcounting ourselves, and we avoid throwing any exceptions in the data layer.

Reply:
That said, there's clearly a problem with my `CSR._deallocate` here - it's probably redundant, because the only situation you _should_ be able to get a `CSR` that doesn't own its data is to have a `scipy` backing.  I think the reason it exists is to do with the `__init__` method possibly raising an exception, but I'm not 100% certain.

The objects are all designed under the assumption that the only other possible owner of data is numpy, but I think we should make that more rigorous.  I really don't think we should be handling any ownership concerns ourselves beyond "I personally allocated this pointer, so I will also deallocate it" - you can see how easy it is to make small mistakes, whereas numpy is battle-tested.

Reply:
If I've done anything similar in currently existing code, we should change it as well - I don't _think_ I did, but if you do spot anything let me know.

Reply:
I am not sure I am following you here.
If a CRS layer has `_deallocate = False` and `_scipy = None`, wouldn't `as_scipy()` break the reference counting anyway?
So the exception is not the issue, but the user having access to those object.

I never intended user using these function and had them `cpdef` for my own debugging...
I hid them and let the user access only to the copy version. The error should them only appear when we mess up in development but never to the user.

Reply:
Yeah, that's why I mean there's a redundancy in `CSR` - it shouldn't be possible to obtain an object with `_deallocate = False` and `_scipy = None`, so having a system where such an invalid object can exist was bad programming on my part.

I think the functions could be accessible to Python space - it's a useful operation that can safely exist with the scipy views.  I suppose it doesn't matter so much if it's `cdef` for now to hide it and we make it visible later with the correct safety procedures. I don't really like the idea of committing something that's a bit fragile, even if it's only intended for devs - I don't like `csr.fast_from_scipy` for a similar reason, really - but if it's compile-only it's probably ok. I should change the unpickle methods to put the proper error checking in, though, since they're absolutely required to be Python accessible.

Reply:
I'm worried that this reference stealing feels deeply thread-unsafe, but that's probably not a major concern - I put a little thought into it in the rest of the data layer, but not enough to be sure that all (non-stateful) operations are thread-safe.  Obviously the scatter/gather pattern in `CSR` is thread unsafe, but that's by definition a stateful operation.